README.md
[![ebay logo](docs/static/logo-vert.png)](http://www.ebayenterprise.com/)

## Magento Couchbase Session Driver (CRUD complaint)

The eBay Enterprise Couchbase Session Driver provides a SessionHandler interface compat w/ PHP 5.5 and greater to interface and execute through Couchbase.

## Contents

1. [Contributing to This Project](CONTRIBUTING.md)
2. [Change Log](CHANGELOG.md)
3. [Setup Guide](INSTALL.md)

## License

Licensed under the terms of the Open Software License v. 3.0 (OSL-3.0). See [LICENSE.md](LICENSE.md) or http://opensource.org/licenses/OSL-3.0 for the full text of the license.

- - -
Copyright © 2015 eBay Enterprise, Inc.


CHANGELOG.md
[![ebay logo](docs/static/logo-vert.png)](http://www.ebayenterprise.com/)

# Change Log
All notable changes to this project will be documented in this file.

## [1.0.0] - 2015-12-04
### CREATED
Initial Public Release For Couchbase Session Driver - Compat. w/ Couchbase Server 3.1.0 and Couchbase Server 4.0.0. SessionHandler Interface for PHP > 5.5, fully compliant. Use latest 1.0.X revision for initial release... IE ~dev-1.0.X in composer.

[1.0.0]: https://github.com/eBayEnterprise/magento-couchbase-backend/tree/1.0.5


src/app/code/local/Mage/Core/Model/Session/Abstract/CouchbaseSession.php
<?php
/**
 * Copyright (c) 2015-2016 eBay Enterprise, Inc.
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.md.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 *
 * DISCLAIMER
 *
 * Do not edit or add to this file if you wish to upgrade Magento to newer
 * versions in the future. If you wish to customize Magento for your
 * needs please refer to http://www.magento.com for more information.
 *
 * @category    Mage
 * @package     Mage_Core 
 * @copyright   Copyright (c) 2015-2016 Ebay Enterprise, Inc. (http://www.ebayenterprise.com)
 * @author      Robert Krule <rkrule@ebay.com>
 * @createdate  December 2, 2015
 * @license     http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 */

class Mage_Core_Model_Session_Abstract_CouchbaseSession implements SessionHandlerInterface {
   /* Bots get shorter session lifetimes */
   /* TODO: Make this a configurable option in ADMIN and auto-detect bots -rkrule */
    const BOT_REGEX          = '/^alexa|^blitz\.io|bot|^browsermob|crawl|^curl|^facebookexternalhit|feed|google web preview|^ia_archiver|indexer|^java|jakarta|^libwww-perl|^load impact|^magespeedtest|monitor|^Mozilla$|nagios|^\.net|^pinterest|postrank|slurp|spider|uptime|crawler|find|NodeUptime|org_bot|QuerySeekSpider|GoogleBot|dnshealthcheckbot|UNDEFINED|bingbot|CCBot|^wget|yandex/i';
	
    const DEFAULT_FIRST_LIFETIME        = 600;      /* The session lifetime for non-bots on the first write */
    const DEFAULT_BOT_FIRST_LIFETIME    = 60;       /* The session lifetime for bots on the first write */

    const DEFAULT_LIFETIME		= 28880;    /* Default Lifetime after first write */
    const DEFAULT_BOT_LIFETIME          = 7200;     /* Lifetime for bots - shorter to prevent bots from wasting backend storage */

    const SLEEP_TIME         		= 500000;   /* Sleep 0.5 seconds between lock attempts (1,000,000 == 1 second) */
    const FAIL_AFTER         		= 15;       /* Try to break lock for at most this many seconds */

   /**
    * Holds the Couchbase connection.
    */
    protected $_connection = null;
 
    /**
    * The Couchbase host
    */
    protected $_host = null;

    /**
    * The Couchbase Cluster Username
    */
    protected $_user = null;

    /**
    * The Couchbase Cluster Password
    */
    protected $_password = null;

    /**
    * The couchbase port
 
    /**
    * The Couchbase bucket name.
    */
    protected $_bucket = null;
 
    /**
    * Define a expiration time by type of user-agent.
    */
    protected $_expire = null;

    /**
    * Set the default configuration params on init.
    */
    public function __construct( $host = 'couchbase://127.0.0.1?config_cache=/tmp/phpcb_cache&http_poolsize=10', $user = 'admin', $password = 'Welcome123', $bucket = 'session') {
        $this->_host = $host;
	$this->_user = $user;
	$this->_password = $password;
        $this->_bucket = $bucket;
    }
 
    /**
    * Open the connection to Couchbase (called by PHP on `session_start()`)
    */
    public function open($savePath, $sessionName) {
        $cluster = new CouchbaseCluster($this->_host, $this->_user, $this->_password);
        $this->_connection = $cluster->openBucket($this->_bucket);
	$this->_connection->__set('operationTimeout', 2000000); // 100000 = 1 second
	return $this->_connection ? true : false;
    }
 
    /**
    * Close the connection. Called by PHP when the script ends.
    */
    public function close() {
        unset($this->_connection);
        return true;
    }
 
    /**
    * Read data from the session.
    */
    public function read($sessionId) {
        $key = $sessionId;

	$result = null;

	//Spinlock on read several times....

	while(true)
	{
		try
		{
      			$result = $this->_connection->get($key)->value;
			break;
		} catch( CouchbaseException $e) {
			if( strpos( $e->getMessage(), "The key does not exist on the server") === false )
			{
				if( $try === self::FAIL_AFTER ) 
                        	{
				      $result = false;
                        	      break;
                       	}

				Mage::Log("In Try: ". $try . " attempting to obtain read for session key: ". $key . " sleeping: ". self::SLEEP_TIME . "(ms)");

                      		usleep(self::SLEEP_TIME);
                    		$try++; 
                  		continue;
			} else {
				//Key doesn't exist, just fail
				$result = false;
				break;
			}
		};
	}

        return $result;    
    }
 
    /**
    * Write data to the session.
    */
    public function write($sessionId, $sessionData) {
        $key = $sessionId;
	$keys = array( $key );

        if(empty($sessionData)) {
            return false;
        }

	$new = null;
	$try = 0;

	try
	{
		$new = $this->_connection->get($key);
	} catch( CouchbaseException $e) {
		//Do nothing....
	};

	if ( $this->_expire === null )
	{
		$userAgent = $_SERVER['HTTP_USER_AGENT'];
		
		if( strcmp( $userAgent, '') !== 0 )
		{
                	$isBot = ! $userAgent || preg_match(self::BOT_REGEX, $userAgent);
                	if ($isBot) {
				if ( $new == null ) // First Write
				{
					$this->_expire = self::DEFAULT_BOT_FIRST_LIFETIME;
					Mage::Log("First Write for Bot detected for user agent: ". $userAgent);
				} else { 
					$this->_expire = self::DEFAULT_BOT_LIFETIME;
                        		Mage::Log("Bot detected for user agent: ". $userAgent);
				}
                	} else {
				if ( $new == null ) // First Write Non Bot
				{
					$this->_expire = self::DEFAULT_FIRST_LIFETIME;
				} else {
					$this->_expire = self::DEFAULT_LIFETIME;
				}
			}
		} else {
			if ( $new == null ) // First Write Non Bot
                        {
                        	$this->_expire = self::DEFAULT_FIRST_LIFETIME;
                        } else {
                                $this->_expire = self::DEFAULT_LIFETIME;
                        }
		}
	}

	try
	{
        	$this->_connection->upsert($key, $sessionData, array('expiry' => $this->_expire ));
		$result = true;
	} catch( CouchbaseException $e ) {
		$result = false;
	};

        return $result;
    }

    /**
    * Delete data from the session.
    */
    public function destroy($sessionId) {
        $keys = array( $sessionId );
	$result = true;

	try
	{
        	$result = $this->_connection->remove($keys);
		$result = true;
        } catch( CouchbaseException $e) {
                $result = false;
        };

        return $result;
    }
 
    /**
    * Run the garbage collection.
    */
    public function gc($maxLifetime) {
        return true;
    }
}
?>


src/app/code/local/Mage/Core/Model/Session/Abstract/Varien.php
<?php
/**
 * Copyright (c) 2015-2016 eBay Enterprise, Inc.
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.md.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 *
 * DISCLAIMER
 *
 * Do not edit or add to this file if you wish to upgrade Magento to newer
 * versions in the future. If you wish to customize Magento for your
 * needs please refer to http://www.magento.com for more information.
 *
 * @category    Mage
 * @package     Mage_Core
 * @copyright   Copyright (c) 2015-2016 Ebay Enterprise, Inc. (http://www.ebayenterprise.com)
 * @author      Robert Krule <rkrule@ebay.com>
 * @createdate  December 2, 2015
 * @license     http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 */

class Mage_Core_Model_Session_Abstract_Varien extends Varien_Object
{
    const VALIDATOR_KEY                         = '_session_validator_data';
    const VALIDATOR_HTTP_USER_AGENT_KEY         = 'http_user_agent';
    const VALIDATOR_HTTP_X_FORVARDED_FOR_KEY    = 'http_x_forwarded_for';
    const VALIDATOR_HTTP_VIA_KEY                = 'http_via';
    const VALIDATOR_REMOTE_ADDR_KEY             = 'remote_addr';
    const SECURE_COOKIE_CHECK_KEY               = '_secure_cookie_check';

    /**
     * Map of session enabled hosts
     * @example array('host.name' => true)
     * @var array
     */
    protected $_sessionHosts = array();

    /**
     * Configure and start session
     *
     * @param string $sessionName
     * @return Mage_Core_Model_Session_Abstract_Varien
     */
    public function start($sessionName=null)
    {
        if (isset($_SESSION) && !$this->getSkipEmptySessionCheck()) {
            return $this;
        }

        // getSessionSaveMethod has to return correct version of handler in any case
        $moduleName = $this->getSessionSaveMethod();
        switch ($moduleName) {
            /**
             * backward compatibility with db argument (option is @deprecated after 1.12.0.2)
             */
            case 'db':
                $moduleName = 'user';
                /* @var $sessionResource Mage_Core_Model_Resource_Session */
                $sessionResource = Mage::getResourceSingleton('core/session');
                $sessionResource->setSaveHandler();
		session_module_name($moduleName);
                break;
            case 'user':
                // getSessionSavePath represents static function for custom session handler setup
                call_user_func($this->getSessionSavePath());
		session_module_name($moduleName);
                break;
            case 'files':
                //don't change path if it's not writable
                if (!is_writable($this->getSessionSavePath())) {
                    break;
                }
		session_module_name($moduleName);
		break;
	    case 'couchbase':
		$config_node = Mage::getConfig()->getNode('global/session_save_couchbase');
        	$connection_string = (string)$config_node->consoleConnectionString;
        	$user = (string)$config_node->consoleUser;
        	$password = (string)$config_node->consolePassword;
		$sessionBucket = (string)$config_node->consoleSessionBucket;

		$handler = new Mage_Core_Model_Session_Abstract_CouchbaseSession($connection_string, $user, $password, $sessionBucket);
		session_set_save_handler($handler, true);
		break;
            default:
                session_save_path($this->getSessionSavePath());
		session_module_name($moduleName);
                break;
        }

        $cookie = $this->getCookie();
        if (Mage::app()->getStore()->isAdmin()) {
            $sessionMaxLifetime = Mage_Core_Model_Resource_Session::SEESION_MAX_COOKIE_LIFETIME;
            $adminSessionLifetime = (int)Mage::getStoreConfig('admin/security/session_cookie_lifetime');
            if ($adminSessionLifetime > $sessionMaxLifetime) {
                $adminSessionLifetime = $sessionMaxLifetime;
            }
            if ($adminSessionLifetime > 60) {
                $cookie->setLifetime($adminSessionLifetime);
            }
        }

        // session cookie params
        $cookieParams = array(
            'lifetime' => $cookie->getLifetime(),
            'path'     => $cookie->getPath(),
            'domain'   => $cookie->getConfigDomain(),
            'secure'   => $cookie->isSecure(),
            'httponly' => $cookie->getHttponly()
        );

        if (!$cookieParams['httponly']) {
            unset($cookieParams['httponly']);
            if (!$cookieParams['secure']) {
                unset($cookieParams['secure']);
                if (!$cookieParams['domain']) {
                    unset($cookieParams['domain']);
                }
            }
        }

        if (isset($cookieParams['domain'])) {
            $cookieParams['domain'] = $cookie->getDomain();
        }

        call_user_func_array('session_set_cookie_params', $cookieParams);

        if (!empty($sessionName)) {
            $this->setSessionName($sessionName);
        }

        // potential custom logic for session id (ex. switching between hosts)
        $this->setSessionId();

        Varien_Profiler::start(__METHOD__.'/start');
        $sessionCacheLimiter = Mage::getConfig()->getNode('global/session_cache_limiter');
        if ($sessionCacheLimiter) {
            session_cache_limiter((string)$sessionCacheLimiter);
        }

        session_start();

        if (Mage::app()->getFrontController()->getRequest()->isSecure() && empty($cookieParams['secure'])) {
            // secure cookie check to prevent MITM attack
            $secureCookieName = $sessionName . '_cid';
            if (isset($_SESSION[self::SECURE_COOKIE_CHECK_KEY])
                && $_SESSION[self::SECURE_COOKIE_CHECK_KEY] !== md5($cookie->get($secureCookieName))
            ) {
                session_regenerate_id(false);
                $sessionHosts = $this->getSessionHosts();
                $currentCookieDomain = $cookie->getDomain();
                foreach (array_keys($sessionHosts) as $host) {
                    // Delete cookies with the same name for parent domains
                    if (strpos($currentCookieDomain, $host) > 0) {
                        $cookie->delete($this->getSessionName(), null, $host);
                    }
                }
                $_SESSION = array();
            }
            if (!isset($_SESSION[self::SECURE_COOKIE_CHECK_KEY])) {
                $checkId = Mage::helper('core')->getRandomString(16);
                $cookie->set($secureCookieName, $checkId, null, null, null, true);
                $_SESSION[self::SECURE_COOKIE_CHECK_KEY] = md5($checkId);
            }
        }

        /**
        * Renew cookie expiration time if session id did not change
        */
        if ($cookie->get(session_name()) == $this->getSessionId()) {
            $cookie->renew(session_name());
        }
        Varien_Profiler::stop(__METHOD__.'/start');

        return $this;
    }

    /**
     * Get session hosts
     *
     * @return array
     */
    public function getSessionHosts()
    {
        return $this->_sessionHosts;
    }

    /**
     * Set session hosts
     *
     * @param array $hosts
     * @return Mage_Core_Model_Session_Abstract_Varien
     */
    public function setSessionHosts(array $hosts)
    {
        $this->_sessionHosts = $hosts;
        return $this;
    }

    /**
     * Retrieve cookie object
     *
     * @return Mage_Core_Model_Cookie
     */
    public function getCookie()
    {
        return Mage::getSingleton('core/cookie');
    }

    /**
     * Revalidate cookie
     * @deprecated after 1.4 cookie renew moved to session start method
     * @return Mage_Core_Model_Session_Abstract_Varien
     */
    public function revalidateCookie()
    {
        return $this;
    }

    /**
     * Init session with namespace
     *
     * @param string $namespace
     * @param string $sessionName
     * @return Mage_Core_Model_Session_Abstract_Varien
     */
    public function init($namespace, $sessionName=null)
    {
        if (!isset($_SESSION)) {
            $this->start($sessionName);
        }
        if (!isset($_SESSION[$namespace])) {
            $_SESSION[$namespace] = array();
        }

        $this->_data = &$_SESSION[$namespace];

        $this->validate();
        $this->revalidateCookie();

        return $this;
    }

    /**
     * Additional get data with clear mode
     *
     * @param string $key
     * @param bool $clear
     * @return mixed
     */
    public function getData($key='', $clear = false)
    {
        $data = parent::getData($key);
        if ($clear && isset($this->_data[$key])) {
            unset($this->_data[$key]);
        }

        return $data;
    }

    /**
     * Retrieve session Id
     *
     * @return string
     */
    public function getSessionId()
    {
        return session_id();
    }

    /**
     * Set custom session id
     *
     * @param string $id
     * @return Mage_Core_Model_Session_Abstract_Varien
     */
    public function setSessionId($id=null)
    {
        if (!is_null($id) && preg_match('#^[0-9a-zA-Z,-]+$#', $id)) {
            session_id($id);
        }
        return $this;
    }

    /**
     * Retrieve session name
     *
     * @return string
     */
    public function getSessionName()
    {
        return session_name();
    }

    /**
     * Set session name
     *
     * @param string $name
     * @return Mage_Core_Model_Session_Abstract_Varien
     */
    public function setSessionName($name)
    {
        session_name($name);
        return $this;
    }

    /**
     * Unset all data
     *
     * @return Mage_Core_Model_Session_Abstract_Varien
     */
    public function unsetAll()
    {
        $this->unsetData();
        return $this;
    }

    /**
     * Alias for unsetAll
     *
     * @return Mage_Core_Model_Session_Abstract_Varien
     */
    public function clear()
    {
        return $this->unsetAll();
    }

    /**
     * Retrieve session save method
     * Default files
     *
     * @return string
     */
    public function getSessionSaveMethod()
    {
        return 'files';
    }

    /**
     * Get sesssion save path
     *
     * @return string
     */
    public function getSessionSavePath()
    {
        return Mage::getBaseDir('session');
    }

    /**
     * Use REMOTE_ADDR in validator key
     *
     * @return bool
     */
    public function useValidateRemoteAddr()
    {
        return true;
    }

    /**
     * Use HTTP_VIA in validator key
     *
     * @return bool
     */
    public function useValidateHttpVia()
    {
        return true;
    }

    /**
     * Use HTTP_X_FORWARDED_FOR in validator key
     *
     * @return bool
     */
    public function useValidateHttpXForwardedFor()
    {
        return true;
    }

    /**
     * Use HTTP_USER_AGENT in validator key
     *
     * @return bool
     */
    public function useValidateHttpUserAgent()
    {
        return true;
    }

    /**
     * Retrieve skip User Agent validation strings (Flash etc)
     *
     * @return array
     */
    public function getValidateHttpUserAgentSkip()
    {
        return array();
    }

    /**
     * Validate session
     *
     * @param string $namespace
     * @return Mage_Core_Model_Session_Abstract_Varien
     */
    public function validate()
    {
        if (!isset($this->_data[self::VALIDATOR_KEY])) {
            $this->_data[self::VALIDATOR_KEY] = $this->getValidatorData();
        }
        else {
            if (!$this->_validate()) {
                $this->getCookie()->delete(session_name());
                // throw core session exception
                throw new Mage_Core_Model_Session_Exception('');
            }
        }

        return $this;
    }

    /**
     * Validate data
     *
     * @return bool
     */
    protected function _validate()
    {
        $sessionData = $this->_data[self::VALIDATOR_KEY];
        $validatorData = $this->getValidatorData();

        if ($this->useValidateRemoteAddr()
                && $sessionData[self::VALIDATOR_REMOTE_ADDR_KEY] != $validatorData[self::VALIDATOR_REMOTE_ADDR_KEY]) {
            return false;
        }
        if ($this->useValidateHttpVia()
                && $sessionData[self::VALIDATOR_HTTP_VIA_KEY] != $validatorData[self::VALIDATOR_HTTP_VIA_KEY]) {
            return false;
        }

        $sessionValidateHttpXForwardedForKey = $sessionData[self::VALIDATOR_HTTP_X_FORVARDED_FOR_KEY];
        $validatorValidateHttpXForwardedForKey = $validatorData[self::VALIDATOR_HTTP_X_FORVARDED_FOR_KEY];
        if ($this->useValidateHttpXForwardedFor()
            && $sessionValidateHttpXForwardedForKey != $validatorValidateHttpXForwardedForKey ) {
            return false;
        }
        if ($this->useValidateHttpUserAgent()
            && $sessionData[self::VALIDATOR_HTTP_USER_AGENT_KEY] != $validatorData[self::VALIDATOR_HTTP_USER_AGENT_KEY]
        ) {
            $userAgentValidated = $this->getValidateHttpUserAgentSkip();
            foreach ($userAgentValidated as $agent) {
                if (preg_match('/' . $agent . '/iu', $validatorData[self::VALIDATOR_HTTP_USER_AGENT_KEY])) {
                    return true;
                }
            }
            return false;
        }

        return true;
    }

    /**
     * Retrieve unique user data for validator
     *
     * @return array
     */
    public function getValidatorData()
    {
        $parts = array(
            self::VALIDATOR_REMOTE_ADDR_KEY             => '',
            self::VALIDATOR_HTTP_VIA_KEY                => '',
            self::VALIDATOR_HTTP_X_FORVARDED_FOR_KEY    => '',
            self::VALIDATOR_HTTP_USER_AGENT_KEY         => ''
        );

        // collect ip data
        if (Mage::helper('core/http')->getRemoteAddr()) {
            $parts[self::VALIDATOR_REMOTE_ADDR_KEY] = Mage::helper('core/http')->getRemoteAddr();
        }
        if (isset($_ENV['HTTP_VIA'])) {
            $parts[self::VALIDATOR_HTTP_VIA_KEY] = (string)$_ENV['HTTP_VIA'];
        }
        if (isset($_ENV['HTTP_X_FORWARDED_FOR'])) {
            $parts[self::VALIDATOR_HTTP_X_FORVARDED_FOR_KEY] = (string)$_ENV['HTTP_X_FORWARDED_FOR'];
        }

        // collect user agent data
        if (isset($_SERVER['HTTP_USER_AGENT'])) {
            $parts[self::VALIDATOR_HTTP_USER_AGENT_KEY] = (string)$_SERVER['HTTP_USER_AGENT'];
        }

        return $parts;
    }

    /**
     * Regenerate session Id
     *
     * @return Mage_Core_Model_Session_Abstract_Varien
     */
    public function regenerateSessionId()
    {
        session_regenerate_id(true);
        return $this;
    }
}
